Improve m.fit() functionality

tjof2 opened the pull request:
### Description of the change
- Closes #1202 by supporting Jacobian for bounded least squares fitting.
- ~~As a result, can deprecate `fitter='mpfit'` in favour of `fitter='leastsq'` - since both are wrappers/derived from MINPACK (Fortran), but the `scipy.optimize.least_squares` route has more functionality, including for numerical gradient approximation.~~
- Add informative error messages for missing analytical gradients (closes #1388)
- Support numerical gradient approximation by finite differences where possible
- ~~Parameter error estimation available when `fitter='L-BFGS-B'`~~
- Added Huber loss function to complement least-squares and Poisson MLE
- Additional global optimization methods available if `scipy` version is >=1.2.0
- Clearer warning message for bounded global optimization (closes #1346)
- ~~Option to remove weighting by inverse variance (closes #1627)~~
- Update docstring for EELSModel fit() (closes #2276)
- Added `print_info=True/False` option to `.fit()`, similar to `.decomposition()`. ~~It is `False` by default since it will print *a lot* when using `m.multifit()`.~~ An example is shown below.
- Improved documentation for fitting algorithms.
- Improved tests for fitting algorithms.

### Progress of the PR
- [x] Changes implemented,
- [x] update docstring,
- [x] update user guide,
- [x] add tests,
- [x] ready for review.

### Minimal example
An example that _fails_ has been chosen deliberately to show the output of `print_info=True`:
```python
>>> m.fit(grad="2-point", fitter="L-BFGS-B", bounded=True, print_info=True)
Fit info:
  fitter=L-BFGS-B
  method=ls
  grad=2-point
  bounded=True
Fit result:
      fun: nan
 hess_inv: <6x6 LbfgsInvHessProduct with dtype=float64>
      jac: array([-2.27957549e-06,  8.37888231e+12, -3.67247360e+07, -1.31072000e+05,
        2.30918888e+04,  8.19200000e+03])
  message: b'ABNORMAL_TERMINATION_IN_LNSRCH'
     nfev: 14
      nit: 0
     njev: 2
   status: 2
  success: False
        x: array([4.72696350e+17, 3.84570141e+00, 1.00000000e-01, 1.00000000e+00,
       2.23000000e+03, 1.00000000e+00])
```

tjof2 committed:
Update requirement to SciPy 0.17 (released Feb 2016) and tidy-up multifit

Add Huber loss functions

Initial support of Jacobian for #1202

Fix bounded jacobian + tests #1202

Update helper functions

Additional tests for L-BFGS-B as per #1388

Separate function for inverse variance weighting, as part of #1627

Check deprecation warnings

Speed up test for Model2D by factor of 20x

Update docstrings

Improve model.rst documentation and update fit() docstring

First pass at updating .fit()

Add Huber test functions

Start supporting other gradients

Start supporting parameter standard deviation for non-linear optimizers where available

Add parameter error estimation for L-BFGS-B

Split test files

Tidy test_fitting.py

Update test_model.py

Parametrize fitting

Added tests for differential evolution method

Further test fixes

Added print_info option like s.decomposition()

Checking grad argument

Support different gradient/jacobian methods

Update docs

Update EELS docstring, closes #2276

tjof2 committed:
Correct jac kwarg and add informative error message for #1388

tjof2 committed:
Update tolerances for tests

tjof2 committed:
Missed changes on last commit

tjof2 committed:
Reduce number of test parametrizations

tjof2 committed:
Update docs

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -14,8 +14,12 @@ dimensions i.e. spectra and images respectively. Most of the syntax is
+
+.. note::
+
+    Plotting and analytic gradient-based fitting methods for are not yet
```
remove for?

tjof2 replied:
Done

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
+"Do you want to...":
+
+* Apply bounds to your model parameter values?
+* Use of gradient information to try and improve the fit result?
```
This sentence is not clear to me. What about: "are gradient based algorithms well suited for your model"?

tjof2 replied:
Yep - changed.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
+    is defined, the default behaviour is to perform a weighted least-squares
+    fit using the inverse variance.
+
+    In this scenario, to explicitly disable weighting, pass ``m.fit(weights=None)``.
```
Why would one want to disable weighting if the variance is defined? Also, wouldn't that be equivalent to remove variance from metadata?

francisco-dlp replied:
In the end, I wonder if we should add the ``weights`` keyword at all...

tjof2 replied:
> Why would one want to disable weighting if the variance is defined? Also, wouldn't that be equivalent to remove variance from metadata?

Yes - but for example, say you estimated the Poisson noise, and wanted to compare the results of weighted vs. unweighted fits.

I would say it is easier to just run `m.fit()` and `m.fit(weights=None)` to compare the results, rather than estimating the variance then unsetting it and (potentially) having to set it again.

francisco-dlp replied:
I see, but the question is: why would you want to compare one with the other when one way is right and the other is not? Some may want to do for, say, educational purposes, but I don't think that it is worth adding a feature for that since that minority can always simply delete the variance metadata.

tjof2 replied:
Sure. I'll tidy it up

tjof2 replied:
Just to be clear here - in removing the `weights` kwarg @francisco-dlp, are you happy for #1627 to just be closed as it is? Perhaps with some clearer documentation that once the variance is set, all fits are weighted until it is unset?

francisco-dlp replied:
I agree that it is a good idea to close #1627 by just documenting how to do it.

tjof2 replied:
Great thanks - I will do that.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
 
 .. code-block:: python
 
    >>> m.fit(fitter="Nelder-Mead", method="ml")
```
Now that you are at it, you could maybe consider deprecating ``ml`` in favour of ``ml-poisson``.

tjof2 replied:
Done

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -849,9 +902,12 @@ model:
-    >>> m.fit(fitter='TNC', method='custom', min_function=my_custom_function)
+    >>> # We must use a general non-linear optimizer again
+    >>> m.fit(fitter='Nelder-Mead',
+    ...       method="custom",
```
Wouldn't it be possible to pass the function directly instead of ``method="custom"`` plus the function in ``min_function``? We could rename ``method``->``loss-function``.

tjof2 replied:
Agreed.

francisco-dlp commented on hyperspy/models/eelsmodel.py:
```diff
@@ -287,49 +286,125 @@ def resolve_fine_structure(
+                * All of the available methods for :py:func:`scipy.optimize.minimize`
+                  can be used here:
+
+                    - "Powell"
```
I think that it is better not to list the methods here because in time this list will be incorrect.

tjof2 replied:
Agreed - done.

francisco-dlp commented on hyperspy/models/model1d.py:
```diff
@@ -554,12 +555,24 @@ def _jacobian(self, param, y, weights=None):
+                                raise ValueError(
+                                    f"Analytical gradient not available for {parameter}. "
+                                    "Consider using a numerical approximation for the gradient "
+                                    "instead by calling `m.fit(grad='2-point')`."
```
by calling **e.g.** ...

And maybe calling ``m.fit(grad=False)`` instead?

tjof2 replied:
Yep changed

francisco-dlp commented on hyperspy/models/model1d.py:
```diff
@@ -554,12 +555,24 @@ def _jacobian(self, param, y, weights=None):
+                                        raise ValueError(
+                                            f"Analytical gradient not available for {par}. "
+                                            "Consider using a numerical approximation for the gradient "
+                                            "instead by calling `m.fit(grad='2-point')`."
```
as above

tjof2 replied:
Done

francisco-dlp commented on hyperspy/models/model1d.py:
```diff
@@ -568,9 +581,21 @@ def _jacobian(self, param, y, weights=None):
                         for parameter in component.free_parameters:
+                            if not callable(parameter.grad):
+                                raise ValueError(
+                                    f"Analytical gradient not available for {parameter}. "
```
Since the error message is the same at multiple places, what about storing it in the global variable instead?

tjof2 replied:
Yep easy to resolve - done.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+
+                    cost = self.fit_output.fun
+                    hess = np.diag(self.fit_output.hess_inv.todense())
+                    self.p_std = np.sqrt(max(1.0, np.abs(cost)) * ftol * hess)
```
This seems to consistently underestimate the error bounds. Besides, that doesn't compute the std, what may be misleading. Therefore I suggest to remove it.
Alternatives that can work for all optimizers would be to use the F-test like [lmfit does](https://lmfit.github.io/lmfit-py/confidence.html#method-used-for-calculating-confidence-intervals), bootstrapping or MLMC with emcee, again like [lmfit](https://lmfit.github.io/lmfit-py/fitting.html#label-emcee).

tjof2 replied:
The alternative is to just use `hess_inv.todense()` as an estimate of the covariance matrix and use the same procedure as `leastsq` - rather than do `np.sqrt(max(1.0, np.abs(cost)) * ftol * hess)`. But that's not particularly clean etc. So I'll remove it as you suggest. 

Taking the lmfit approach (or another) is probably best left for a separate PR.

francisco-dlp commented on hyperspy/models/eelsmodel.py:
```diff
@@ -287,49 +286,125 @@ def resolve_fine_structure(
+                * All of the available methods for :py:func:`scipy.optimize.minimize`
+                  can be used here:
+
+                    - "Powell"
```
as above

tjof2 replied:
Done

francisco-dlp commented on hyperspy/models/eelsmodel.py:
```diff
@@ -287,49 +286,125 @@ def resolve_fine_structure(
+        kind="std",
+        **kwargs,
+    ):
+        """Fits the model to the experimental data.
```
Instead of copying the docstring it would be better to put it in a variable like e.g. in https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/signal.py#L2339

tjof2 replied:
Done.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+                  data. Also known as Poisson maximum likelihood estimation
+                  (MLE). Not available if ``fitter`` is ``"leastsq"``
+                  or ``"odr"``.
+                * "huber" minimize the Huber loss function. Not available
```
The huber loss-function takes the important ``delta`` parameter. Shouldn't we offer a way to input the parameter? The usual way would be to add an extra keyword. An alternative could be passing e.g. ``"huber delta=10"``.

tjof2 replied:
Good catch - will resolve.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -727,44 +755,48 @@ def _set_p0(self):
                                else self.p0 + parameter.value)
 
-    def set_boundaries(self):
+    def set_boundaries(self, bounded=True):
```
The ``bounded`` keyword is not documented

tjof2 replied:
Done

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -727,44 +755,48 @@ def _set_p0(self):
                                else self.p0 + parameter.value)
 
-    def set_boundaries(self):
+    def set_boundaries(self, bounded=True):
```
Maybe we could consider making this method private? (With just a deprecation warning now of course.)

tjof2 replied:
Agreed - will add.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+
+        if pcov is None:  # Indeterminate covariance
+            p_var = np.zeros(len(self.p0), dtype=float)
+            p_var.fill(np.inf)
```
Why not filling with ``nan`` instead?

tjof2 replied:
This is consistent with how `scipy.optimize.curve_fit` does it. I can replace it with `np.nan` if you prefer, but I think `np.inf` is fine.

francisco-dlp replied:
The things is that in hyperspy several functions (e.g. ``print_summary_statistics``, 2D plotting constrast adjustement...) know how to deal with ``nans`` but not with ``inf``s.

tjof2 replied:
OK, I'll replace with nan in that case.

ericpre replied:
Would it be better to fix the functions which doesn't support `np.inf` to keep the consistency with scipy?

francisco-dlp replied:
I don't know how to best do that. Actually, I don't know why ``curve_fit`` uses ``inf`` instead of ``nan`` for this, since a ``nan`` seems to me a better description of the situation. In hyperspy: if a dataset contains ``inf``s, I think that it is up to the user to decide how to best handle them, because the best approach depends on the situation.

tjof2 replied:
To be fair I think nan is probably better here after some though - it's not that the parameter estimate is very (i.e. infinitely) large, rather, it cannot be estimated accurately and therefore should not be relied upon.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+                  bounds on parameters. See :py:mod:`scipy.odr` for more details.
+                * All of the available methods for :py:func:`scipy.optimize.minimize`
+                  can be used here:
+
```
As there is not reliable way to keep this list in sync with scipy, I suggest to remove it.

tjof2 replied:
Agreed - done.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
+    | ``"Differential Evolution"`` |  Yes   | No        | No     | All            | global |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+
+\* ``mpfit`` is deprecated since HyperSpy 1.6. Use ``"leastsq"`` instead.
```
I remember testing them a while ago with a few datasets and finding that mpfit gave better results. However, my tests weren't very systematic. Have you tested the performance of mpfit vs leastsq with bounds?

tjof2 replied:
Hmmm, that'll probably be because `mpfit` uses the Levenberg–Marquardt algorithm with its own implementation of bounds, while `leastsq` with bounds actually uses the Trust Region Reflective algorithm (`method="trf"`).

There are a couple of options here, both of which could preserve `mpfit`.

1. Deprecate the options `["leastsq", "mpfit"]` with e.g. `["lm", "trf", "dogbox"]` with clearer documentation.
2. Don't do anything, leave `"mpfit"` as it is.

Option (1) is nice because the default fitter for `m.fit()` could then be `None`, and we could smartly suggest an appropriate method depending on the type of function, bounds, and gradients.

Option (2) is OK, but I think it would be better to deprecate `mpfit` for `lm` (Levenberg–Marquardt).

ericpre replied:
Since mpfit is vendored and it is not supported since a while, at some point, it may break. If we don't have a good reason to keep mpfit, we should deprecate it.

tjof2 replied:
With a little bit of work it can probably support passing a custom minimizer (like we did for `s.decomposition()` and sklearn objects), so you could ultimately use `lmfit`, `scikit-optimize` or `openopt/copt` if desired. That could be option 3.

francisco-dlp replied:
That sounds excellent, but maybe something for a different PR? A possible problem (for which solutions exist) with that approach is that it won't work for parallel distributed fits since the library will need to be present in all the nodes.

Regarding mpfit, as mentioned above, in the past, I have obtained better results with it than with bounded ``scipy.optimize.least_squares``, so, unless we show that this not actually generally the case, I suggest that we keep it unless it gets broken beyond easy repair at some point. it seems that a maintained version of the code can be found in https://github.com/segasai/astrolibpy . However, it is not distributed in pypi as far as I can tell. Ours is also a maintained version of the code and so-far it has been easy to fix it every time that changes in numpy broke something. For the future (i.e. after The Split) If we want to continue offering mpfit, I suggest to contact @segasai to check if he would be happy to publish his code (that is the original code from which we forked our version a long ago) in pypi and conda-forge.

I like @tjof2's 1. option. Currently it seems that even for unbounded optimization the default bounding method is "trf", while it would be more efficient to use "lm". An ``auto`` mode could fix this, while leaving the user the option to choose a different method if required. If we keep mpfit, we could add ``mpfit`` to the list of bounds methods. Also, the optimizer in that case could be "LM" instead of "leastsq" as it is now, since "leastsq" is rather the loss-funtion and we could use all other optimizers to solve the same problem.

tjof2 replied:
Thanks @francisco-dlp  - I also like option 1. I'll work on that.

ericpre replied:
OK, maybe it would be worth trying to sort out the distribution of mpfit or even better integrate it in scipy!

francisco-dlp replied:
That's a good point: if the presumed advantage comes from a particular way of implementing the bounds, maybe that can be added to scipy's ``least_squares`` as an extra bounds method.

tjof2 replied:
I think there is a (small) bug in the calculation of `m.p_std` for mpfit, it treats the `perror` variable from mpfit as being the estimated covariance matrix, when it isn't - its the estimated parameter standard deviations (prior to multiplying by any cost function). So if your assertion "mpfit gave better results" is based purely off the values in `p_std`, it might not necessarily be better. It's easy to fix though and may not necessarily be a problem.

tjof2 replied:
I've found a nicer way to keep mpfit rather than deprecating (for now)

tjof2 replied:
> Since mpfit is vendored and it is not supported since a while, at some point, it may break. If we don't have a good reason to keep mpfit, we should deprecate it

The code is actually much simpler than it looks @ericpre. It probably looks bad due to all the old IDL comments, old Fortran comments and old Python style (or lack of...). I've actually tidied it up a lot, I can easily push it in a separate PR. Since it's written in pure Python (unlike `scipy.optimize.least_squares` which wraps Fortran), it can be a little slower when optimizing, but the bound constraints are useful, so keeping it (for now at least) makes sense.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
+
+* Apply bounds to your model parameter values?
+* Use of gradient information to try and improve the fit result?
+* Estimate the errors of the parameter values found by the fit?
```
errors -> std or confidence interval?

tjof2 replied:
Yep - std is correct here.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
+* Use of gradient information to try and improve the fit result?
+* Estimate the errors of the parameter values found by the fit?
+* Fit your data in the least-squares sense, or is your data corrupted
+  with Poisson noise?
```
when the data is corrupted by Poisson noise it is still possible to successfully use weighted least squares as long as the number of counts is not too low. This has the advantage that the std and the red-chisq can be computed.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -742,29 +743,44 @@ estimated standard deviation are stored in the following line attributes:
+Optimization algorithms that take into account the gradient of the minimization function
+often out-perform so-called "derivative-free" optimization algorithms in terms of how rapidly
+they converge to a solution. HyperSpy can use analytical gradients for model-fitting, as well
+as numerical estimates of the gradient based on finite differences. Following the above example:
```
I would also mention here that:

* by default, the optimizers use analytical gradient and that using ``grad=True`` (as in your example) turns on analytical gradient if available for all components in the model.
* the advantages of analytical gradients are that i) they're more accurate ii) the optimization runs faster since there is no need to estimate the gradient

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -742,29 +743,44 @@ estimated standard deviation are stored in the following line attributes:
+    provided. See :ref:`signal.noise_properties` for more information.
 
+Using gradient information
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I would move this subsection under WLS below.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
    >>> line.coefficients.std
    (0.0055752036447948173, 0.46950832982673557)
 
+Instead of using the inverse variance, you can also provide your own
```
This is equivalent to setting the variance to the inverse of those weights. From The Zen of Python:

> There should be one-- and preferably only one --obvious way to do it.

tjof2 replied:
Yep!

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
+Maximum likelihood estimation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To avoid biased estimation in the case of data corrupted by Poisson noise,
```
I would add here "with very few counts" as it is only then when this becomes advantageous.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
-We can use Poisson maximum likelihood estimation
-instead, which is an unbiased estimator for poissonian noise.
-To do so, we use a general optimizer called "Nelder-Mead".
+Maximum likelihood estimation
```
Poisson ML estimation?

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
+biased. A more accurate result can be obtained with weighted least squares,
+where the weights are proportional to the inverse of the noise variance.
+Although this is still biased for Poisson noise, it is a good approximation
+in most cases.
```
in most cases -> when there is a sufficient number of counts.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
    (1.0004224896604759, -0.46982916592391377)
    >>> line.coefficients.std
    (0.0055752036447948173, 0.46950832982673557)
 
```
It would be good to explain here that the variance can be set directly in metadata. Actually, probably it would be good to add a ``set_variance`` method that takes a hyperspy signal or np array and set the metadata entry for convinience.

It would be good to mention that the variance is only uses with ``method=="ls"``

tjof2 replied:
I've added a `set_noise_variance()` method for convenience.

> It would be good to mention that the variance is only uses with `method=="ls"`

Is this true? Currently it says:
```
    weights = None

    if weights == "inverse_variance":
        weights = self._convert_variance_to_weights()

    args = (self.signal()[np.where(self.channel_switches)], weights)
```
and then `args` is passed to the optimizer, be it `leastsq`, `odr`, `mpfit` or the general `minimize()`

francisco-dlp replied:
In that case, I think that there is a bug in there: the weights should only be applied with method is LS to turn it into WLS. If we apply the weights when using Poisson-ML we get wrong results. Applying weights probably makes sense with the Huber-loss function though...

tjof2 replied:
I'll tidy it up!

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
+    supported by HyperSpy. The options listed above are generally sufficient for most applications,
+    but for more information on the other supported local and global optimization algorithms,
+    please see the `SciPy documentation <http://docs.scipy.org/doc/scipy/reference/optimize.html>`_.
 
```
What about adding a paragraph here to explain that HyperSpy supports several loss functions, including the custom one, and the pass to give details of each one. Maybe it'll be good to put all the subsections below (except for the gradient one)  under a Fitting methods or Fitting Loss Functions subsections for clarity.

tjof2 replied:
Yep - working on that

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +791,80 @@ gaussian noise and proceed to fit as in the previous example.
+Custom optimizations
+^^^^^^^^^^^^^^^^^^^^
+
+As well as the built-in ``ls``, ``ml`` and ``huber`` optimization functions,
```
It would be good to mention that currently analytical gradients are not supported for custom loss-functions.

tjof2 replied:
Sure.

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"Nelder-Mead"``            |  No    | No        | No     | All            | local  |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"L-BFGS-B"``               |  Yes   | Yes       | Yes    | All            | local  |
```
If we decide to remove the conf limits estimation for L-BFGS-B as I suggest below, Errors should be back to ``False`` here.

tjof2 replied:
Yep - done.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+
+    def fit(
+        self,
+        fitter="leastsq",
```
What about deprecating ``fitter`` in favour of ``optimizer``?

tjof2 replied:
Sure!

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+                  gradient argument ``min_function_grad``. Not available
+                  if ``fitter`` is ``"leastsq"`` or ``"odr"``.
+
+        grad : bool or ["2-point", "3-point", "cs"] or callable, default False
```
As it was, ``grad`` was a switch to use the analytical gradient. When ``False`` the numerical gradient was computed internally for the optimizers that use the gradient. It think that now it is not clear what happens when ``grad`` is ``False``.

tjof2 replied:
I've got a few ideas for this - I'll test them out.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -727,44 +755,48 @@ def _set_p0(self):
-                        self.mpfit_parinfo.extend((
-                            {'limited': limited,
-                             'limits': limits},) * param._number_of_elements)
+    def set_mpfit_parameters_info(self, bounded=True):
```
Maybe this one could be privatized too?

tjof2 replied:
Agreed - done

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+            _logger.warning(
+                "Covariance of the parameters could not be estimated. "
+                "Estimated parameter standard deviations will be np.inf. "
+                "This could indicate that the model is a poor description "
```
Is that always true? If not, I suggest to remove it not to worry the users unnecessarily.

tjof2 replied:
Well, the use of "could" qualifies it, but I agree, it could be an unnecessary warning in that sense - removed.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+        method="ls",
+        grad=False,
+        bounded=False,
+        ext_bounding=False,
```
That that you are at it, I think that we can deprecate ``ext_bounding``, this dates from the times when there was no optimizer with bound support available and is a (very) poor man's solution to the problem.

tjof2 replied:
Sure - have done so.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+                  :py:func:`scipy.optimize.shgo` for more details on available
+                  options. Requires ``scipy >= 1.2.0``.
+
+        method : {"ls", "ml", "huber", "custom"}, default "ls"
```
As suggested above, what about ``method`` -> ``loss_function``?

tjof2 replied:
Agreed - done.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+                If the array has the same dimensions as the signal,
+                and ``method="ls"``, then weighted least squares
+                is performed using the array values as the weights.
+        print_info : bool, default False
```
Wouldn't it be better a ``return_info`` keyword that returns a ``OptimizeResult`` object with a nice ``__repr__`` method or even a subclass of it with html printing? In this way the user can more easily reuse that data.

For multifit this could return a similear ``OptimzeResult`` object with suitable HyperSpy signals as attributes for easy inspection of the result (the current print, as you point out, is quite useless for multi-dimensional fitting).

tjof2 replied:
I'll take a look at seeing if `self.fit_output` can be standardized into an `OptimizeResult` object for the other fitting methods.

tjof2 replied:
I've added a `return_info` argument and it will always return an `OptimizeResult` object for all fitting methods.

francisco-dlp replied:
What would be best, to store i. this info in ``fit_output`` as present ii. return it optionally, replacing ``print_info`` as I suggested above and deprecate ``fit_output``?

Actually, ``fit_output`` is currently not stored anywhere when saving the model, what I think is a bug. If we take option i, then we won't need ``print_info`` or ``return_info`` in ``multifit``, but maybe we could consider a ``store_info`` keyword to make it optional. Also we should consider storing this info when saving the model. In my opinion option 1 is better because it is consistent with the fact that we also store other fitting info, such as the chisq, as model attributes.

tjof2 replied:
Ok, so let's store it and not return it. I think the print is a useful option still.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+        Notes
+        -----
+        The chi-squared and reduced chi-squared statistics, and the
+        degrees of freedom, are computed automatically when fitting.
```
Maybe it'll be good to add here that they are computed only when fitting using least-squares.

tjof2 replied:
Yes good idea - done.

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
-                          "fmin_l_bfgs_b": "L-BFGS-B",
-                          "fmin_tnc": "TNC",
-                          "fmin_powell": "Powell"}
+        optimizer_dict = {
```
Maybe ``deprecated_optimizer_dict`` for clarity?

tjof2 replied:
Done

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
+
+                        jacobian = _wrap_jac
+
+                    elif grad is False or grad is None:
```
Then maybe make ``"2-point"`` the default and deprecate ``None`` and ``False``?

tjof2 replied:
Yep.

francisco-dlp reviewed (changes requested):
This is a much needed improvement of the fitting documentation and code. I have left quite a few comments, mainly with some extra suggestions now that you are at it.

tjof2 commented:
Thanks @francisco-dlp for the detailed review - will work through it over the weekend!

tjof2 commented:
Thanks @francisco-dlp for the detailed review - will work through it over the weekend!

tjof2 committed:
Initial changes as per review from @francisco-dlp

Minor edit

tjof2 committed:
Small corrections to docs

tjof2 committed:
Add .set_noise_variance() and .get_noise_variance() helpers

tjof2 committed:
Deprecate 'fitter=' in favour of 'optimizer='

tjof2 commented:
Last commit was a bit of a mess - will update tomorrow

tjof2 committed:
Move docstring to separate file

Deprecate 'method' for 'loss_function'

Update EELS docstring

Add check for analytical gradients

Improvements to model.fit()

Revert accidental style format

Update broken tests

Fix gradient methods and docstring, remove weights argument as discussed

Tidy docstring and table

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into improve-fit

tjof2 committed:
Deprecate return_info

tjof2 committed:
Logger warning if optimization failed for any reason (e.g. convergence)

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into improve-fit

tjof2 committed:
Update documentation orders and testing

tjof2 committed:
Improve testing of fit functions for Model1D

tjof2 committed:
Additional testing for scipy optimize result

tjof2 committed:
Set global tolerance

tjof2 committed:
Support Huber delta

tjof2 committed:
Update testing and docstring

Add tests for helpers

Test Model2D better and add catch for ODR + Model2D

Fix bug in helpers

Update documentation

tjof2 commented:
I've addressed most of the comments now @francisco-dlp. 

Here are a few outstanding things which I personally think are better tackled in separate PRs to ensure they are done properly.

1. Storing and saving `model.fit_output`.
2. Concatenating/grouping/storing `model.fit_output` when calling `model.multifit()` - it could potentially get quite large for lots of pixels so we may only want to store some of the info
3. Using `grad="analytical"` by default if all the components implement it. This *changes* the default behaviour and hence the fitting results compared to current/previous HyperSpy versions (evidenced by a whole bunch of test failures elsewhere if I try it), whereas this PR only deprecates the old choice of `grad=True/False` without changing the end results for now. Doing it in a separate PR would be cleaner to make sure the change is correct, and it could be profiled to make sure e.g. it's not a performance regression.

Here, I've used `grad="auto"` to mean auto-differentiation where available. Perhaps a better keyword is `grad="fd"` for finite-difference, and preserve `"auto"` for item (3)? That ties in better with the kwarg `fd_scheme="2-point"` etc.

tjof2 committed:
Revert accidental formatting

tjof2 committed:
Fix test broken in previous commit

thomasaarholt commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +653,88 @@ Fitting the model to the data
+the question "Do you want to...":
+
+* Apply bounds to your model parameter values?
+* Are gradient-based fitting algorithms appropriate for your model?
```
"Do you want to..." "are gradient-based fitting..." doesn't really work. 
Maybe: "Use gradient-based fitting algorithms?"

thomasaarholt commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +653,88 @@ Fitting the model to the data
+    supported by HyperSpy; however, only some of them support bounds and/or gradients. For more information,
+    please see the `SciPy documentation <http://docs.scipy.org/doc/scipy/reference/optimize.html>`_.
+
+    \*\* Requires ``scipy >= 1.2.0``.
```
Scipy 1.5.2 is out, and I notice that hyperspy [setup.py currently requires scipy 0.15](https://github.com/hyperspy/hyperspy/blob/fe7602e4ff9ae51a86c2a4c0d0545971ef972ae3/setup.py#L48) (released 2015). Maybe we can just set scipy 1.2 as required version? I notice conda installs scipy 1.5.0.

tjof2 replied:
I have no objection, I was conservative in advancing it to 0.17.0 here. If @francisco-dlp and @ericpre have no objections/reasons why this is a bad idea, then we can advance it further

ericpre replied:
Ubuntu 18.04 LTS have [scipy 0.19.1](https://packages.ubuntu.com/search?suite=default&section=all&arch=any&keywords=python3-scipy). We are not dealing dependencies requirement consistently but since there is no significant advantages to change it, I would keep it as broad as possible. Also, I suspect that scipy is one our dependencies which is the most difficult to build.

tjof2 replied:
I'll keep it at 0.17

thomasaarholt commented on doc/user_guide/model.rst:
```diff
@@ -659,65 +653,88 @@ Fitting the model to the data
+^^^^^^^^^^^^^^
+
+HyperSpy supports a number of loss functions. The default is ``"ls"``,
+or the least-squares loss. For the vast majority of cases, this loss
```
Remove the "or" -> "The default is ``"ls"``, the least-squares loss."

thomasaarholt reviewed (commented):
Docs look great, only some minor things.

tjof2 committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into improve-fit

tjof2 committed:
Update docs (thanks @thomasaarholt)

tjof2 committed:
Fix merge conflict

tjof2 committed:
Fix merge conflict

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -917,7 +1010,7 @@ component using mpfit and bounds on the ``centre`` parameter.
     >>> g1.centre.bmin = 7
     >>> g1.centre.bmax = 14
     >>> g1.centre.bounded = True
```
I think that we can remove this line, since the ``bounded`` attribute is now gone.

tjof2 replied:
Done

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -775,43 +790,77 @@ gaussian noise and proceed to fit as in the previous example.
+To avoid biased estimation in the case of data corrupted by Poisson noise
+with very few counts, we can use Poisson maximum likelihood estimation (MLE) instead.
+This is an unbiased estimator for Poisson noise. To perform MLE, we must
+use a general, non-linear optimizer from the ref:`table above <optimizers-table>`,
```
this link to the table is broken.

tjof2 replied:
Done

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -884,26 +935,68 @@ passed, using the following signature:
+solution. HyperSpy can use analytical gradients for model-fitting,
+as well as numerical estimates of the gradient based on finite differences.
+
+If all the components in the model support analytical gradients,
```
I think that we need a ``versionadded`` directive here.

tjof2 replied:
Done

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -924,475 +1010,672 @@ def red_chisq(self):
-                    method)
-            args = (self.signal()[np.where(self.channel_switches)],
-                    weights)
+                    "Weighted fitting is not supported for `loss_function='ml_poisson'`. "
```
I think that in this case, it is better to perform the operation without weights. Indeed a typical workflow is:

* Get close to the optimal parameters using WLS
* Refine with ML-Poisson.

It would be rather inconvenient having to delete the variance to perform these operations.

tjof2 replied:
Ok, perhaps change this to a logger warning (i.e. that the noise variance is set but weighted fitting won't actually be done), and set `weights = None` to proceed then?

tjof2 replied:
Done.

francisco-dlp replied:
For once I think that doing nothing is better: users don't like warnings and this one seems unnecessary to me: somebody using ML-Poisson should know what he's doing...

tjof2 replied:
OK will do!

francisco-dlp commented on doc/user_guide/model.rst:
```diff
@@ -928,17 +1021,60 @@ component using mpfit and bounds on the ``centre`` parameter.
+
+After fitting the model, details about the optimization
+procedure, including whether it finished successfully,
+are stored in the attribute ``fit_output`` as a
```
In the future, we may add the feature of storing the multi-dimensional output instead of the last fit output in ``fit_output``. Planning for that, what would you think about replacing ``print_info`` with ``return_info``? I bet that it'll print the info all the same, with the added advantage that it makes it easier to explore the output of the latest fit.

tjof2 replied:
I agree in principle. What about for consistency:

- `s.decomposition()` has `print_info=True/False` and `return_info=True/False` (implemented in #2336 and #2383 )
- `m.fit()` has `print_info=True/False` and `return_info=True/False` (implemented here?)

Note that `print_info` as it currently stands does also highlight a few of the key options you passed to `m.fit()`, so for example if in future we went with automatically choosing the `optimizer` or `grad`, you could print that out. While `m.fit_output` is simply an `OptimizeResult` object at the moment, with no additional details.

tjof2 replied:
I agree in principle. What about for consistency:

- `s.decomposition()` has `print_info=True/False` and `return_info=True/False` (implemented in #2336 and #2383 )
- `m.fit()` has `print_info=True/False` and `return_info=True/False` (implemented here?)

Note that `print_info` as it currently stands does also highlight a few of the key options you passed to `m.fit()`, so for example if in future we went with automatically choosing the `optimizer` or `grad`, you could print that out. While `m.fit_output` is simply an `OptimizeResult` object at the moment, with no additional details.

For example:
```
    Fit info:
      optimizer=L-BFGS-B
      loss_function=ls
      bounded=False
      grad="fd"
```

The printing also currently drops these two attributes of `OptimizeResult`
```
            # Drop these as they can be large (== size of data array)
            output_print.pop("fun", None)
            output_print.pop("jac", None)
```
Which if you have a large array, ends up printing a lot of output. But they can be returned of course.

tjof2 replied:
`print_info` could also be used to control the various `verbose/quiet` options of the different optimizers (in future)

tjof2 replied:
As it stands, I've just pushed a commit that implements `return_info`.

francisco-dlp commented on hyperspy/docstrings/model.py:
```diff
@@ -0,0 +1,99 @@
+            as part of the optimization. This parameter has no effect if
+            ``optimizer`` is a derivative-free or global optimization method.
+
+                * "auto" uses a finite difference scheme (if available) for numerical
```
I think that it'll be clearer to rename `"auto"` -> `"fd"`

tjof2 replied:
Done

francisco-dlp commented on hyperspy/docstrings/model.py:
```diff
@@ -0,0 +1,99 @@
+FIT_PARAMETERS_ARG = """optimizer : str or None, default None
+            The optimization algorithm used to perform the fitting.
+
+                * "lm" performs least-squares optimization using the
```
Although I see the point of dividing each bounding approach into separate optimizers, I think that this can lead to conceptual confusion. Why not using ``"lm"`` for the optimizer and then use ``bounded`` to choose between ``"trf", "dogbox" and "mpfit"?

By the way, you've made mpfit the default optimizer for bounded-LM, is that because you've performed some tests and you've found it to be superior to the others?

tjof2 replied:
But "trf" and "dogbox" aren't the Levenberg-Marquardt algorithm, so that doesn't make sense.

> mpfit the default optimizer for bounded-LM

It's the _only_ bounded Levenberg-Marquardt option.

"trf" and "dogbox" are alternative least-squares minimization algorithms, which do support bounds and (optionally) sparse Jacobians. But they are not the Levenberg-Marquardt algorithm (`optimizer="lm"`).

The approach taken here is the closest one for preserving backwards-compatibility/replicating previous behaviour.

francisco-dlp replied:
You are right, scipy's API, where you get those options in ``method`` (what I (mis)understood as bounding_method), led me to conceptual misunderstanding, so clearly your approach is clearer :)

francisco-dlp reviewed (changes requested):
This looks almost ready to me, except for the couple of things that I mention below.

tjof2 committed:
Minor doc fixes, and change grad='auto' to grad='fd' as per comments from @francisco-dlp

tjof2 committed:
Logger warning about weights when ml-poisson is being used, but proceed with unweighted fit as per @francisco-dlp comment

tjof2 committed:
typo in last commit

tjof2 committed:
Remove warning message

tjof2 committed:
Support return_info from m.fit()

tjof2 committed:
typo
